### PR TITLE
[WIP] shell: get rid of the proxy

### PR DESCRIPTION
**Pull request by fatih**

### CODE REVIEW

File: internal/cmd/shell/shell.go [hunk 1]
```diff
@@ -157,8 +174,12 @@ second argument:
 				"root",
 				"-s",
 				"-t", // the -s (silent) flag disables tabular output, re-enable it.
-				"-h", host,
-				"-P", port,
+				"-h", cert.AccessHost,
+				"-P", strconv.Itoa(cert.Ports.MySQL),
+				"--ssl-mode", "VERIFY_IDENTITY",
+				"--ssl-ca", "/etc/ssl/cert.pem", // change depending on OS
```
[the remark this answers is not in the document]
Comment (mdlayher) [reply internal/cmd/shell/shell.go:180]: I did a bit of digging and found the following for Linux, in addition to whatever is specified via environment variables: https://cs.opensource.google/go/go/+/refs/tags/go1.18:src/crypto/x509/root_linux.go

```
const (
	// certFileEnv is the environment variable which identifies where to locate
	// the SSL certificate file. If set this overrides the system default.
	certFileEnv = "SSL_CERT_FILE"

	// certDirEnv is the environment variable which identifies which directory
	// to check for SSL certificate files. If set this overrides the system default.
	// It is a colon separated list of directories.
	// See https://www.openssl.org/docs/man1.0.2/man1/c_rehash.html.
	certDirEnv = "SSL_CERT_DIR"
)
```

This seems pretty complicated and ultimately may merit either filing a Go proposal to get access to the locations Go considers for CA cert lookup, or using a third party package which copies out this sort of functionality from the standard library.

File: internal/cmd/shell/shell.go [hunk 2]
```diff
@@ -138,16 +119,52 @@ second argument:
...
+			err = pem.Encode(keyPem, &pem.Block{
+				Type:  "PRIVATE KEY",
+				Bytes: privBytes,
+			})
+			if err != nil {
+				return err
+			}
+
+			clientKeyPath := filepath.Join(os.TempDir(), "client-key.pem")
+			err = os.WriteFile(clientKeyPath, keyPem.Bytes(), 0600)
```
[the remark this answers is not in the document]
Comment (mdlayher) [reply internal/cmd/shell/shell.go:142]: From what I can tell, I don't see any options for passing certificates via stdin: https://dev.mysql.com/doc/mysql-shell/8.0/en/mysql-shell-encrypted-connections.html.

I do wonder if it's possible to have some sort of ephemeral "file" owned by this process, which can be quickly destroyed on process exit, but also has a name and can be referenced as long as the parent process is alive.

I'll have to dig around to see if there is some kind of Linux syscall that could do the job.

[the remark this answers is not in the document]
Comment (mdlayher) [reply internal/cmd/shell/shell.go:142]: A viewer on Twitch suggests that `/dev/stdin` would work, but then we also have the problem of needing to specify both cert and key. They also suggest `<(echo "cert data")` which would use shell magic to create a temporary file descriptor. I wonder if we could make that work at least on *nix.

**Review by 0xETHengineer — APPROVED** [review]

**Comment by mscoutermarsh** [comment]
Closing as stale 😄

**Pull request closed without merging**